src/CargoBuild/CargoBuildTask.cs

```diff
@@ -0,0 +1,437 @@
+
+                Log.LogMessage(MessageImportance.Normal, $"CargoBuild, Auth Enabled: {EnableAuth}");
+
+                foreach (var projects in rustProjects)
+                {
+                    string path = projects;
+
+                    var fetchTask = RustFetchAsync(path, EnableAuth);
+                    tasks.Add(fetchTask);
+                }
+
+                await Tasks.Task.WhenAll(tasks.ToArray());
+                bool success = tasks.Select(x => (int)x.Result).Sum() == 0;
+                stopwatch.Stop();
+                if (success)
+                {
+                    Log.LogMessage(MessageImportance.Normal, $"---- CargoBuild fetching Completed Successfully in {stopwatch.Elapsed.Seconds} seconds ----\n\n");
+                }
+                else
+                {
+                    Log.LogError($"---- CargoBuild fetching had an issue. Check the build log for details. ----\n\n");
+                }
+
+                return success;
+            }
+            catch (Exception ex)
+            {
+                LogException(ex);
+                return false;
+
+                void LogException(Exception ex)
+                {
+                    if (ex is AggregateException aggEx)
+                    {
+                        foreach (Exception innerEx in aggEx.InnerExceptions)
+                        {
+                            LogException(innerEx);
+                        }
+                    }
+                    else
+                    {
+                        Log.LogErrorFromException(ex, showStackTrace: true);
+                    }
+                }
+            }
+            finally
+            {
+                if (_shouldCleanRustPath)
+                {
+                    CleanupRustPath();
+                }
+            }
+        }
+
+        private async Task<ExitCode> RustFetchAsync(string workingDir, bool authorize = false)
+        {
+            ExitCode authResult = authorize ? await DoRegistryAuthAsync(workingDir) : ExitCode.Succeeded;
+
+            if (authorize && authResult == ExitCode.Succeeded || !authorize)
+            {
+                string path;
+                string args;
+                const int RetryCnt = 2;
+
+                path = _cargoPath;
+                args = "fetch";
+
+                Log.LogMessage(MessageImportance.Normal, $"Fetching cargo crates for project in {workingDir}");
+                var exitCode = await ExecuteWithRetriesAsync(path, processArgs: args, workingDir, retryCount: RetryCnt, processRetryArgs: args);
+                Log.LogMessage(MessageImportance.Normal, $"Finished fetching cargo crates for project in  {workingDir}");
+                return exitCode;
+            }
+
+            return authResult;
+        }
+
+        private async Task<ExitCode> ExecuteWithRetriesAsync(string processFileName, string processArgs, string workingDir, int retryCount, string? processRetryArgs = null)
```

Comment:
What're the retries for?

Other SDKs don't do the same thing. Like for a csproj if you have a syntax error it doesn't just try again a bunch of times.

Are we expecting inherent flakiness with cargo?